MAINT: Fix B006 violations across astropy

MridulS opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request is to address and fix #16107

There are probably places that this may actually break intended functionality, I have added the `noqa` label where I think it's breaking the expected behavior. But in general we should avoid creating mutable arguments in function definitions.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

mhvk commented on astropy/units/function/core.py:
```diff
@@ -193,7 +195,7 @@ def physical_type(self):
         return self.physical_unit.physical_type
 
-    def is_equivalent(self, other, equivalencies=[]):
+    def is_equivalent(self, other, equivalencies=None):
```
This is wrong - like for regular `is_equivalent`, `None` has the meaning at no equivalencies at all are used.

mhvk commented on astropy/units/core.py:
```diff
@@ -2431,7 +2460,9 @@ def __copy__(self):
 
-    def decompose(self, bases=set()):
+    def decompose(self, bases=None):
+        if bases is None:
```
I think we should not just change things to do exactly the same thing but just use more cycles. Here, the next line clearly is checking whether `bases` is set, so can be `if bases is None and ...`. Similarly the `if len(bases) == 0` at the end could be worked into here.

mhvk commented on astropy/units/core.py:
```diff
@@ -2244,7 +2266,9 @@ def represents(self):
 
-    def decompose(self, bases=set()):
+    def decompose(self, bases=None):
+        if bases is None:
```
This kind of change seems illogical - one should just adjust `decompose` to also take `None`.

mhvk commented on astropy/units/core.py:
```diff
@@ -1942,7 +1962,9 @@ def represents(self):
-    def decompose(self, bases=set()):
+    def decompose(self, bases=None):
+        if bases is None:
+            bases = set()
```
This is not needed, one would just do `if bases is not None and self not in bases:` below.

eerovaher commented on astropy/coordinates/earth.py:
```diff
@@ -903,11 +905,13 @@ def __len__(self):
         # Conversion to another unit in both ``to`` and ``to_value`` goes
         # via this routine. To make the regular quantity routines work, we
         # temporarily turn the structured array into a regular one.
+        if equivalencies is None:
+            equivalencies = []
         array_view = self.view(self._array_dtype, np.ndarray)
         if equivalencies == []:
             equivalencies = self._equivalencies
```
```suggestion
        array_view = self.view(self._array_dtype, np.ndarray)
        if equivalencies is None:
            equivalencies = self._equivalencies
```

mhvk commented on astropy/time/core.py:
```diff
@@ -3179,6 +3179,8 @@ def to(self, unit, equivalencies=[]):
         to_value : get the numerical value in a given unit.
         """
+        if equivalencies is None:
+            equivalencies = []
```
No, this is wrong: just like for `units`, `equivalencies=None` has concrete meaning, so this changes API.

eerovaher commented on astropy/coordinates/spectral_quantity.py:
```diff
@@ -226,6 +226,8 @@ def to(self, unit, equivalencies=[], doppler_rest=None, doppler_convention=None)
             New spectral coordinate object with data converted to the new unit.
         """
         # Make sure units can be passed as strings
+        if equivalencies is None:
```
This is changing how the code works.

mhvk replied:
Again, this changes meaning.

eerovaher commented on astropy/coordinates/transformations/graph.py:
```diff
@@ -483,6 +483,8 @@ def to_dot_graph(
         """
         # We store nodes as `dict` keys because differently from `list` uniqueness is
         # guaranteed and differently from `set` insertion order is preserved.
+        if addnodes is None:
+            addnodes = []
```
`addnodes` is only used in a single line, so it's better to have the `if addnodes is None` check just before that line.

eerovaher reviewed (changes requested):
I had a quick look at `coordinates`.

mhvk commented on astropy/nddata/tests/test_decorators.py:
```diff
@@ -333,7 +333,9 @@ def test_parameter_default_identical_to_explicit_passed_argument():
     # should still raise a Warning and use the explicit one.
     @support_nddata
-    def func(data, meta={"a": 1}):
+    def func(data, meta=None):
```
This breaks the purpose of the test!

mhvk commented on astropy/modeling/utils.py:
```diff
@@ -286,7 +286,9 @@ class _SpecialOperatorsDict(UserDict):
 
-    def __init__(self, unique_id=0, special_operators={}):
+    def __init__(self, unique_id=0, special_operators=None):
+        if special_operators is None:
```
Here, `super()` should be adjusted rather than initialize here.

mhvk commented on astropy/modeling/polynomial.py:
```diff
@@ -1909,13 +1917,17 @@ def __init__(
+            ap_coeff = {}
         self._ap_order = ap_order
         self._bp_order = bp_order
         self._ap_coeff = ap_coeff
```
If one went ahead, this should be `ap_coeff if ap_coeff is not None else {}`

mhvk reviewed (changes requested):
I've glanced through a bit but looked more carefully at units, and do not think this PR as it stands should go through. In general, avoiding mutable defaults seems good, but not if all one ends up doing is just changing the code to initialize the same default right after: that is more cycles for no benefit (for working code). Instead, one would need to go through carefully and adjust the code to make use of the new default. I added some inline comments, but stress that I stopped because it was clear that for units at least, this is not the right approach.

Overall, my suggestion would be *not* to enforce the rule, but just change those places where it has no performance impact and where `None` is a fine alternative. 

But if others feel this should be enforced for some modules, then please do exclude `units` - I do not want to have `noqa` put in the (many) places where `equivalencies=[]` is used (and really would prefer that also not to be done in `coordinates`, `time` and `table`).

namurphy commented:
I'm very much in favor of enforcing ruff rule [B006](https://docs.astral.sh/ruff/rules/mutable-argument-default/), so thank you for submitting this!  Here's my go-to example for why mutable default arguments should be avoided:

```python
>>> def f(l=[]):
...     l.append(len(l))
...     return l
... 
>>> f()
[0]
>>> f()
[0, 1]
```

Even if code is working now, there's a chance that when changing code, mutable default arguments can lead to weird behavior that can take a while to track down.  This weird behavior can sometimes happen in edge cases that we forget to test.

To exclude certain subpackages from the rule, it would be possible to set up [per-file ignores](https://docs.astral.sh/ruff/settings/#lint_per-file-ignores) in the ruff configuration.  That way, it wouldn't be necessary to have `# noqa` comments.  (My personal preference would be to put `# noqa: B006` comments directly in the code, since they would serve as extra motivation to avoid mutable default arguments.)

If the mutability of lists isn't necessary within a function, changing the default argument to an empty tuple may be a viable option.  Tuples are generally more efficient than lists, so this may lead to a very slight performance improvement.

Since `astropy.units` would require special attention, it seems like it would be worth separating those changes off from this pull request.  I often try to split up pull requests like this into one pull request per subpackage.

Thank you again!

mhvk commented:
Sadly, `equivalencies=[]` is not used only in units, and indeed because it is used elsewhere, this PR introduced several changes to `equivalencies=None` which, if I had looked at units only, would have led to rarely-exposed, hard-to-find, subtle bugs. 

Unfortunately, an empty tuple for `equivalencies` would have different meaning. In principle, we could switch to something like `np._NoValue`, I guess, but I still feel that then we'd be letting what is admittedly a generally good rule define code in ways that are simply not necessary.

I'll add that really my problem with this rule-imposing stuff is that it requires time to do it right, since one has to think about the meaning of the code. And, sadly, that means time from the reviewers rather than from the submitters, since only the subpackage reviewers understand the code well enough. Over the last months, I have probably spent about a whole work week reviewing these things, finding problems in quite a few cases, and honestly I do not think this is very good use of my time.

MridulS commented:
I agree with @namurphy and @mhvk both. These violations can easily introduce subtle bugs (the whole reason I started going this way was in the search of memory leaks in MaskedColumn) but this PR in particular isn't really helping with cleaning up the code properly.

I'll close this draft for now, and hopefully create a new one which addresses some of these issues in the better way.
